File: TrainingFolder2019/20190309 MSU Trinity 2014/b.py
from decimal import Decimal, getcontext

getcontext().prec = 50

from math import sin,cos,tan,pi

K = int(input())

alcohol = pi / 2 / K
ang = [(alcohol * i) for i in range(K+K)]
_sin = [sin(ang[i]) for i in range(K+K)]
_cos = [cos(ang[i]) for i in range(K+K)]
_cot = [1/tan(ang[i]) if i > 0 else -1 for i in range(K+K)]

pos = [[] for i in range(K+K)]

for alpha in range(1,K+K):
	for beta in range(1,K+K-alpha):
		for gamma in range(alpha+1, K+K-beta):
			x =  _sin[alpha] * _sin[beta] / _sin[alpha + beta] * (_cot[alpha] - _cot[gamma])
			pos[gamma].append(x)

for gamma in range(K+K):
	pos[gamma].sort()

eps = 1e-12

import bisect

ans = 0
for i in range(1,K+K):
	j = K+K-i
	for x in pos[i]:
		#for y in pos[j]:
		#	if abs(x + y - 1) < eps:
		#		ans += 1
		#continue
		l = bisect.bisect_left(pos[j], 1-x-eps)
		r = bisect.bisect_right(pos[j], 1-x+eps)
		for y in pos[j][l:r]:
			if abs(x + y - 1) < eps:
				ans += 1

print(ans)

File: TrainingFolder2019/20190309 MSU Trinity 2014/b.cpp
#include <bits/stdc++.h>

using namespace std;

using real_t = long double;

const real_t PI = acosl(-1);

int K;
real_t ang[500], _sin[500], _cos[500], _cot[500];
vector<real_t> pos[500];
vector<pair<real_t, int>> vec[500];

int main() {
	scanf("%d", &K);
	
	real_t alcohol = PI / 2 / K;
	for(int i = 1; i < K + K; i++) {
		ang[i] = alcohol * i;
		_sin[i] = sinl(ang[i]);
		_cos[i] = cosl(ang[i]);
		if(i != K) _cot[i] = 1/tanl(ang[i]);
	}

	for(int alpha = 1; alpha < K + K; alpha++) {
		for(int beta = 1; alpha + beta < K + K; beta++) {
			for(int gamma = alpha + 1; gamma < K + K - beta; gamma++) {
				real_t x = _sin[beta] * _sin[gamma - alpha] / _sin[alpha + beta] / _sin[gamma];
				//printf("%Lf %Lf\n", _sin[beta] * _sin[gamma - alpha] / _sin[alpha + beta], _sin[beta] * _sin[gamma - alpha] / _sin[alpha + beta] / _sin[gamma]);
				pos[gamma].push_back(x);
			}
		}
	}


	const real_t eps = 1e-11;
	for(int gamma = 1; gamma < K+K; gamma++) {
		sort(pos[gamma].begin(), pos[gamma].end());
		for(int i = 0; i < int(pos[gamma].size()); ) {
			int j = i;
			while(j < int(pos[gamma].size()) && fabsl(pos[gamma][i] - pos[gamma][j]) < eps) {
				j += 1;
			}
			vec[gamma].emplace_back(pos[gamma][i], j - i);
			i = j;
		}
	}

	int ans = 0;
	for(int i = 1; i <= K+K; i++) {
		int j = K + K - i; 
		for(auto xi : vec[i]) {
			auto l = lower_bound(vec[j].begin(), vec[j].end(), make_pair(1 - xi.first - eps, 0));
			auto r = upper_bound(vec[j].begin(), vec[j].end(), make_pair(1 - xi.first + eps, 0));
			//assert(r - l <= 5);
			for(auto it = l; it != r; it++) {
				if(fabsl(xi.first + it->first - 1) < eps) ans += it->second * xi.second;
			}
		}
	}

	printf("%d\n", ans);
	return 0;
}